File: movie-ticket-booking-system/backend/src/main/java/com/example/movieticketbooking/service/ShowService.java
package com.example.movieticketbooking.service;

import com.example.movieticketbooking.dto.ShowDto;
import com.example.movieticketbooking.entity.Show;

import java.util.List;

public interface ShowService {
    List<ShowDto> getUpcomingShowsByMovie(Long movieId);
    List<ShowDto> getShowsByMovieAndTheatre(Long movieId, Long theatreId);
    List<ShowDto> getAllShows();
    ShowDto getShowById(Long id);
    Show findEntityById(Long id);
    ShowDto createShow(ShowDto showDto);
    void cancelShow(Long id);
}

File: movie-ticket-booking-system/frontend/src/app/services/show.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { Show } from '../models/show.model';
import { ApiResponse } from '../models/api-response.model';

@Injectable({
  providedIn: 'root'
})
export class ShowService {
  private apiUrl = `${environment.apiUrl}/shows`;

  constructor(private http: HttpClient) {}

  getAllShows(): Observable<ApiResponse<Show[]>> {
    return this.http.get<ApiResponse<Show[]>>(this.apiUrl);
  }

  getShowById(id: number): Observable<ApiResponse<Show>> {
    return this.http.get<ApiResponse<Show>>(`${this.apiUrl}/${id}`);
  }

  getShowsByMovie(movieId: number): Observable<ApiResponse<Show[]>> {
    return this.http.get<ApiResponse<Show[]>>(`${this.apiUrl}/movie/${movieId}`);
  }

  getShowsByMovieAndTheatre(movieId: number, theatreId: number): Observable<ApiResponse<Show[]>> {
    return this.http.get<ApiResponse<Show[]>>(`${this.apiUrl}/movie/${movieId}/theatre/${theatreId}`);
  }

  createShow(show: any): Observable<ApiResponse<Show>> {
    return this.http.post<ApiResponse<Show>>(this.apiUrl, show);
  }

  cancelShow(id: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${id}`);
  }
}
